Atomic pointer operations

Tokazama opened the pull request:
This provides the following public bindings to intrinsic atomic pointer methods:

* `unsafe_load` => `atomic_pointerref`
* `unsafe_store!` => `atomic_pointerset` 
* `unsafe_replace!` => `atomic_pointerreplace`
* `unsafe_swap!` => `atomic_pointerswap`
* `unsafe_modify!` => `atomic_pointermodify`

The arguments follow the same pattern as their field accessor counterparts with the order being the final positional argument. 

The corresponding intrinsic methods don't provide an option to move to the ith address of the pointer so I just looked at what the C code for `pointerref` was doing there and imitated it in Julia, but it would be nice to have someone double check that.

Tokazama committed:
atomic pointer loads and stores

This provides and interface to `Intrinsics.atomic_pointerref` and
`Intrinsics.atomic_pointerset` through `unsafe_load` and
`unsafe_store!`.

Tokazama committed:
Don't change `@atomic`

This was a bad idea so it's removed

Tokazama committed:
Add index argument to load/store

Tokazama committed:
fix docstrings

longemen3000 commented on base/pointer.jl:
```diff
@@ -109,12 +111,19 @@ Incorrect usage may segfault your program or return garbage answers. Unlike C, d
 memory region allocated as different type may be valid provided that the types are compatible.
 """
 unsafe_load(p::Ptr, i::Integer=1) = pointerref(p, Int(i), 1)
+unsafe_laod(p::Ptr, order::Symbol) = atomic_pointerref(p, order)
```
laod?

Tokazama replied:
D'oh

Tokazama committed:
Fix typo

Tokazama committed:
Support for additional atomic ops on pointers

Added the following atomic pointer operations
    unsafe_modify!
    unsafe_replace!
    unsafe_swap!

Tokazama committed:
Export, document and test new methods

Tests previously using the explicit intrinsic method here are replaced
by the new user facing variants

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+"""
+    unsafe_modify!(p::Ptr{T}, op, x, [order::Symbol]) -> Pair
+
+These atomically perform the operations to get and set a field after applying
```
nit: it is not really legal to use this on a field, since those have slightly different semantics, so it should only really be for interop with existing C libraries that require it. It may be helpful also to add an additional statement that this is compatible with C11 and C++11 memory semantics for `_Atomic` and `std::atomic` respectively, per the C++23 standard.
```suggestion
These atomically perform the operations to get and set a memory address after applying
```

The relevant C++23 standard text wording that would be applicable here is:
```
Implementations are recommended to ensure that the representation of
_Atomic(T) in C is same as that of std::atomic<T> in C++ for every
possible type T. The mechanisms used to ensure atomicity and memory
ordering should be compatible.
```

Tokazama replied:
How about this boiler plate

```
The `unsafe` prefix on this function indicates that no validation is performed on the
pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
that referenced memory is not freed or garbage collected while invoking this function.
Incorrect usage may segfault your program. Unlike C, storing memory region allocated as
different type may be valid provided that that the types are compatible.

This is compatible with C11 and C++11 memory semantics for `_Atomic` and `std::atomic`
respectively, per the C++23 standard. Therefore, implementations that atomically
manipulate `Ptr{T}` are recommended to ensure that representation of `_Atomic(T)` in C.
The mechanisms used to ensure atomicity and memory ordering should be compatible.

```

vtjnash replied:
seems like this phrase lots some of the grammatical helper words when I originally wrote it:
"...storing to a memory region allocated as a different type is valid, provided..."
(or `accessing`, or `reading from`, depending on context)

vtjnash replied:
> Therefore, implementations that atomically
manipulate `Ptr{T}` are recommended to ensure that representation of `_Atomic(T)` in C.
The mechanisms used to ensure atomicity and memory ordering should be compatible.

There is only one Julia implementation, so we aren't writing a spec for implementations of it. This should convey that the Julia implementation is compatible with use on pointers declared with `_Atomic` and `std::atomic` type in C11 and C++23 respectively. But not all C types compatible with `_Atomic` are translated at present to work in Julia, so the operation may instead throw a Julia error that the type `T` is not supported.

Tokazama replied:
> There is only one Julia implementation, so we aren't writing a spec for implementations of it.

Sorry, poor use of words on my part. I was trying to find an more elegant way of saying "code that passes `Ptr{T}` to a method with a atomic order argument should ensure ..."

> This should convey that the Julia implementation is compatible with use on pointers declared with _Atomic and std::atomic type in C11 and C++23 respectively. But not all C types compatible with _Atomic are translated at present to work in Julia, so the operation may instead throw a Julia error that the type T is not supported.

I wonder if some of this should start going in "doc/src/base/multi-threading.md" under the "Atomic Operations" heading since its relevant to atomic operations in Julia as a whole and not just these methods. It's also starting to be quite a big chunk of text if we want a light explanation on what these do in the REPL

vtjnash replied:
Yeah, it is probably best to put it in multi-threading.md and cross-reference it. We could also potentially make a kwdoc for it too, so it can be queried from the REPL help (in base/docs/basedocs.jl)

```julia
"""
For unsafe pointer operations, ...
And see also...
"""
kw"atomic"
```

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+    z = op(y, x)
+    unsafe_store!(p, z)
+    return y => z
+
```
nit: add the standard boiler plate warning:
```suggestion

The `unsafe` prefix on this function indicates that no validation is performed on the
pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
that referenced memory is not freed or garbage collected while invoking this function.
Incorrect usage may segfault your program. Unlike C, storing memory region allocated as
different type may be valid provided that that the types are compatible.

```

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+    unsafe_replace!(p::Ptr{T}, expected, desired,
+                  [success_order::Symbol, [fail_order::Symbol=success_order]) -> (; old, success::Bool)
+
+These atomically perform the operations to get and conditionally set a field to
```
same updates here as for `unsafe_modify!`

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+     This function requires at least Julia 1.10.
+"""
+function unsafe_modify!(p::Ptr, op, x, order::Symbol=:not_atomic)
+    Core.Intrinsics.atomic_pointermodify(p, op, x, order)
```
```suggestion
    return atomic_pointermodify(p, op, x, order)
```

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
 unsafe_store!(p::Ptr{T}, x, i::Integer=1) where {T} = pointerset(p, convert(T,x), Int(i), 1)
+unsafe_store!(p::Ptr, x, order::Symbol) = atomic_pointerset(p, x, order)
+function unsafe_store!(p::Ptr, x, i::Integer, order::Symbol)
+    unsafe_store!(p + (aligned_sizeof(eltype(p)) * (Int(i) - 1)), x, order)
```
nit: `aligned_sizeof` isn't generally correct here, since the alignment of the basic type is not necessarily the same as the alignment of an atomic type (generally speaking, it is next-power-of-two instead, up to 2x the register size or 16 bytes). Though currently all of the underlying `Intrinsics.atomic` will error if you try to pass in any type for which `aligned_sizeof` != `Core.sizeof` or is larger than 8 bytes.
```suggestion
    unsafe_store!(p + (Core.sizeof(eltype(p)) * (Int(i) - 1)), x, order)
```

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+!!! compat "Julia 1.10"
+     This function requires at least Julia 1.10.
+"""
+function unsafe_modify!(p::Ptr, op, x, order::Symbol=:not_atomic)
```
I realized looking back at the implementation that `not_atomic` might not be fully implemented here. I think codegen will likely gracefully handle that, but runtime_intrinsics probably won't. We should check that the behavior is consistent for this case.

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+
+These atomically perform the operations to get and conditionally set a field to
+a given value.
+
+    y = unsafe_load(p, fail_order)
+    ok = y === expected
+    if ok
+        unsafe_store!(p, desired, success_order)
+    end
+    return (; old = y, success = ok)
```
I wonder if it would be better to just cross-reference their `•field` equivalents from here for their non-atomic reference (e.g. see `Core.replacefield!`). Thoughts?

Tokazama replied:
Do you think that will be enough context to understand what this does?

vtjnash replied:
Maybe not. It might be better to copy it, so the user doesn't need to cross-reference them.

Tokazama replied:
I kept the verbose example and added refs too.

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+function unsafe_replace!(p::Ptr{T}, expected, desired, success_order::Symbol=:not_atomic, fail_order::Symbol=success_order) where {T}
+    @inline
+    xT = desired isa T ? desired : convert(T, desired)
+    Core.Intrinsics.atomic_pointerreplace(p, expected, xT, success_order, fail_order)
```
```suggestion
    return atomic_pointerreplace(p, expected, xT, success_order, fail_order)
```

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+    Core.Intrinsics.atomic_pointerreplace(p, expected, xT, success_order, fail_order)
+end
+function unsafe_replace!(p::Ptr{Any}, @nospecialize(expected), @nospecialize(desired), success_order::Symbol=:not_atomic, fail_order::Symbol=success_order)
+    Core.Intrinsics.atomic_pointerreplace(p, expected, desired, success_order, fail_order)
```
(I don't think this method is actually needed?)
```suggestion
    return atomic_pointerreplace(p, expected, desired, success_order, fail_order)
```

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+     This function requires at least Julia 1.10.
+"""
+function unsafe_swap!(p::Ptr{Any}, x, order::Symbol=:not_atomic)
+    Core.Intrinsics.atomic_pointerswap(p, x, order)
```
```suggestion
    return atomic_pointerswap(p, x, order)
```
(but similarly, not sure this is needed)

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+function unsafe_swap!(p::Ptr{T}, x, order::Symbol=:not_atomic) where {T}
+    @inline
+    xT = x isa T ? x : convert(T, x)
+    return Core.Intrinsics.atomic_pointerswap(p, xT, order)
```
```suggestion
    return atomic_pointerswap(p, xT, order)
```

vtjnash reviewed (commented):
I think this direction is good. I left a couple comments (mostly just on the documentation)

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Improve docs

Add notes about atomic implementation and refs to relevant field and
property methods

Tokazama committed:
Merge branch 'JuliaLang:master' into atomic_loads

Seelengrab commented on base/pointer.jl:
```diff
@@ -96,34 +96,168 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+
+See also: [`replacefield!`](@ref), [`replaceproperty!`](@ref)
+"""
+function unsafe_replace!(p::Ptr{T}, expected, desired, success_order::Symbol=:not_atomic, fail_order::Symbol=success_order) where {T}
+    @inline
+    xT = desired isa T ? desired : convert(T, desired)
+    return atomic_pointerreplace(p, expected, xT, success_order, fail_order)
+end
```
Are implicit conversions on pointer stores a good idea..?

Seelengrab replied:
Just noticed that regular `unsafe_store!` also does that :pensive:

Tokazama replied:
Would your preference be that it just throws an error otherwise

Seelengrab replied:
Well, if the `convert` doesn't work, the intrinsic checks the type anyway so it wouldn't really change anything error-wise other than making the user choose how to convert in all cases (and possibly giving a different error). I just think it's an unexpected conversion, so I though I'd ask :shrug:

Seelengrab replied:
After some digging, seems like this is from way back when in 2012, in f880afdbc25aacfb6cca970fd451741329fc089f by @vtjnash - it's a style thing I guess, it's just unexpected to have that conversion there when pretty much everything else about `unsafe_*` requires the user to make sure their code is already correct.

Seelengrab replied:
My reasoning is this - JET can't find this, unless you turn on `:sound` mode, which can often give false positives IIRC:

```
julia> struct MyInt
           a::Int
       end

julia> Base.convert(::Type{Int}, a::MyInt) = a

julia> a = [2]
1-element Vector{Int64}:
 2

julia> f(a) = unsafe_store!(pointer(a), MyInt(1))
f (generic function with 1 method)

julia> JET.@report_call f(a)
No errors detected


julia> f(a)
ERROR: TypeError: in pointerset, expected Int64, got a value of type MyInt
Stacktrace:
 [1] unsafe_store!(p::Ptr{Int64}, x::MyInt) (repeats 2 times)
   @ Base ./pointer.jl:126 [inlined]
 [2] f(a::Vector{Int64})
   @ Main ./REPL[10]:1
 [3] top-level scope
   @ REPL[12]:1

julia> JET.@report_call mode=:sound f(a)
═════ 1 possible error found ═════
┌ @ REPL[10]:1 unsafe_store!(pointer(a), MyInt(1))
│┌ @ pointer.jl:126 #self#(p, x, 1)
││┌ @ pointer.jl:126 Base.pointerset(p, convert(T, x), Base.Int(i), 1)
│││ this builtin function call may throw: Base.pointerset(p::Ptr{Int64}, convert(T::Type{Int64}, x::MyInt)::MyInt, Base.Int(i::Int64)::Int64, 1)
││└──────────────────
```

If this type equality were required before calling the intrinsic, these kinds of mistakes would be found by JET without having to make sure the whole codebase is `:sound` :shrug:

Tokazama replied:
Should I change it to this
```julia
unsafe_store!(p::Ptr{T}, x, i::Integer=1) where {T} = unsafe_store!(p, convert(T, x), Int(i))
unsafe_store!(p::Ptr{T}, x::T, i::Int=1) where {T} = pointerset(p, x, i, 1)
```

Seelengrab replied:
I'd personally not have the `convert` at all, but that's not my call to make :shrug:

Tokazama replied:
I'm assuming that would be breaking?

Seelengrab replied:
Technically by SemVer I don't think so (the `convert` is undocumented), but it'd be good to have this discussion some place other than this PR, to not hold it up too much :) As long as the `convert` in your atomic bindings is also undocumented, it should be fine to remove at a later date together with removing it in regular `unsafe_store!`.

If we do want to resolve this with this PR (though that definitely requires more input than just mine here), it'd be good to make it consistent, i.e. such that `unsafe_modify!` also calls `convert` in a similar manner.

Tokazama replied:
Made a new issue here https://github.com/JuliaLang/julia/issues/49826

Seelengrab commented on base/pointer.jl:
```diff
@@ -96,34 +96,168 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+
+"""
+    unsafe_replace!(p::Ptr{T}, expected, desired,
+                  [success_order::Symbol, [fail_order::Symbol=success_order]) -> (; old, success::Bool)
```
```suggestion
                   [success_order::Symbol[, fail_order::Symbol=success_order]]) -> (; old, success::Bool)
```

Minor alignment/formatting :) Looking at `replacefield!`, seems like the second `]` is missing there as well :thinking:

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Sukera <11753998+Seelengrab@users.noreply.github.com>

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

vtjnash commented on base/pointer.jl:
```diff
@@ -96,34 +96,168 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
             p::Ptr{T}, dims::NTuple{N,<:Integer}; own::Bool = false) where {T,N} =
     unsafe_wrap(Atype, p, convert(Tuple{Vararg{Int}}, dims), own = own)
 
+const ATOMIC_PTR_DOCS = """
```
?

vtjnash commented on base/pointer.jl:
```diff
@@ -96,34 +96,168 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
 Incorrect usage may segfault your program or return garbage answers. Unlike C, dereferencing
 memory region allocated as different type may be valid provided that the types are compatible.
+
+This is compatible with loading pointers declared with `_Atomic` and `std::atomic`
```
This clause should probably state that this only applies if `order` is given [, or if order != :not_atomic]. I forgot this is the text for both cases.

We also should try to make a decision whether `order != :not_atomic` should behave exactly the same as `unsafe_load`. Currently internally those may make slightly different alignment and size decisions. In particular, `aligned_sizeof` assumes the object is being stored in an Array. But here we are working with Ptr, and usually `unsafe_load`/`unsafe_store!` ignore mutability and the like and simply load or store bytes for an object of the given size at the given offset and allocate a *new* mutable object with those.

Tokazama replied:
If one doesn't actually spend anytime looking at internal code here I'd bet the assumption is that `:not_atomic` would perform that exact same as the unannotated method variant. Is there an important reason we have `:not_atomic` behave differently?

vtjnash replied:
No, they probably should behave the same as the non atomic versions. There are different behaviors of those however, so the code needs to be updated accordingly to make sure it does that.

Tokazama replied:
Sorry, that's obvious now that I'm reading through these comments again. Should this PR change the generated code and runtime C code for this or solve this in a separate PR?

vtjnash replied:
It should be a separate PR

vtjnash commented on base/pointer.jl:
```diff
@@ -96,34 +96,168 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
 unsafe_load(p::Ptr, i::Integer=1) = pointerref(p, Int(i), 1)
+unsafe_load(p::Ptr, order::Symbol) = atomic_pointerref(p, order)
+function unsafe_load(p::Ptr, i::Integer, order::Symbol)
+    unsafe_load(p + (aligned_sizeof(eltype(p)) * (Int(i) - 1)), order)
```
We need to use a definition of `aligned_sizeof` that actually corresponds to what the intrinsics expect.

```julia
function elsize(::Type{Ptr{T}}) where T
    # this only must return something valid for values which satisfy is_valid_intrinsic_elptr(T),
    # which includes Any and most concrete datatypes
    T === Any && return sizeof(Ptr{Any})
    T isa DataType || sizeof(Any) # throws
    return LLT_ALIGN(Core.sizeof(T), datatype_alignment(T))
end
```

```suggestion
    unsafe_load(p + (elsize(typeof(p)) * (Int(i) - 1)), order)
```

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
delete old pointr doc const

Tokazama committed:
Move general atomic doc strings to kw doc and ref in multi-threading

Tokazama committed:
fix docs

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,146 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
 unsafe_store!(p::Ptr{T}, x, i::Integer=1) where {T} = pointerset(p, convert(T,x), Int(i), 1)
+unsafe_store!(p::Ptr{T}, x, order::Symbol) where {T} = atomic_pointerset(p, x isa T ? x : convert(T,x), order)
+function unsafe_store!(p::Ptr, x, i::Integer, order::Symbol)
+    unsafe_store!(p + (Core.sizeof(eltype(p)) * (Int(i) - 1)), x, order)
```
```suggestion
    unsafe_store!(p + (elsize(typeof(p)) * (Int(i) - 1)), x, order)
```

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,146 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
 
 Load a value of type `T` from the address of the `i`th element (1-indexed) starting at `p`.
-This is equivalent to the C expression `p[i-1]`.
+This is equivalent to the C expression `p[i-1]`. Optionally, an ordering can be provided.
```
Instead of `"an ordering"` can you say that it is the "the atomic memory ordering". Or, more verbosely, to borrow from C++:
> specifies how memory accesses, including regular, non-atomic memory accesses, are to be ordered around an atomic operation

To say that "Optionally, the `order` argument can be used to specify whether other memory access are permitted to be reordered around this operation when they are visible to other threads."

Tokazama committed:
Update base/pointer.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Improve doc string wording

Tokazama committed:
defined `elsize(::Type{<:Ptr})`

Tokazama committed:
Merge branch 'master' into atomic_loads

Tokazama commented on base/pointer.jl:
```diff
@@ -98,32 +98,146 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
 unsafe_load(p::Ptr, i::Integer=1) = pointerref(p, Int(i), 1)
+unsafe_load(p::Ptr, order::Symbol) = atomic_pointerref(p, order)
+function unsafe_load(p::Ptr, i::Integer, order::Symbol)
+    unsafe_load(p + (elsize(typeof(p)) * (Int(i) - 1)), order)
```
I added a definition of for `elsize`, but there's a comment around that method saying it should be deprecated. Would it be better to just replace `elsize(typeof(p))` here with `aligned_sizeof(eltype(p))`?

vtjnash replied:
That comment seems like it is now out of date and should have been removed in 6312bfec433f6ba3d6469f316df4a998be054cfe (#38128)

vtjnash commented on base/array.jl:
```diff
@@ -255,6 +255,7 @@ end
 # Deprecate this, as it seems to have no documented meaning and is unused here,
 # but is frequently accessed in packages
 elsize(@nospecialize _::Type{A}) where {T,A<:Array{T}} = aligned_sizeof(T)
+elsize(@nospecialize T::Type{<:Ptr}) = aligned_sizeof(eltype(T))
```
As I said before, you need to use the correct elsize definition: https://github.com/JuliaLang/julia/pull/49811/files/17540efcbe1d3347c66125c1275dc4e7b755cb61#r1199658397
> We need to use a definition of `aligned_sizeof` that actually corresponds to what the intrinsics expect.
> 
> ```julia
> function elsize(::Type{Ptr{T}}) where T
>     # this only must return something valid for values which satisfy is_valid_intrinsic_elptr(T),
>     # which includes Any and most concrete datatypes
>     T === Any && return sizeof(Ptr{Any})
>     T isa DataType || sizeof(Any) # throws
>     return LLT_ALIGN(Core.sizeof(T), datatype_alignment(T))
> end
> ```

Tokazama committed:
Fix elsize definition

vtjnash committed:
Update base/array.jl

vtjnash commented on base/array.jl:
```diff
@@ -252,9 +252,14 @@ function bitsunionsize(u::Union)
+    T === Any && return sizeof(Ptr{Any})
+    T isa DataType || sizeof(Any) # throws
+    return LLT_ALIGN(Core.sizeof(T), datatype_alignment(T))
+end
```
Can you add a couple tests for `elsize(Ptr{Any})`, `elsize(Ptr{Ref{Any}})`, `elsize(Ptr{NTuple{3,Int8}})`, `elsize(Ptr)`, `elsize(Ptr{Int})`, and `elsize(Ptr{Cvoid})`

Tokazama committed:
Add tests for elsize

Tokazama commented:
I think I messed up doc builds so I'll have to take a look at that a little later before this is merged.

Tokazama committed:
Merge branch 'master' into atomic_loads

Tokazama committed:
fix doc string refs

vtjnash commented on test/intrinsics.jl:
```diff
@@ -107,16 +107,25 @@ end
 const ReplaceType = ccall(:jl_apply_cmpswap_type, Any, (Any,), T) where T
 
+@testset "elsize(::Type{<:Ptr})" begin
+    @test Base.elsize(Ptr{Any}) == 8
+    @test Base.elsize(Ptr{NTuple{3,Int8}}) == 3
+    @test Base.elsize(Ptr{Cvoid}) == 0
+    @test Base.elsize(Ptr{Base.RefValue{Any}}) == 8
+    @test Base.elsize(Ptr{Int}) == 8
```
```suggestion
    @test Base.elsize(Ptr{Any}) == sizeof(Int)
    @test Base.elsize(Ptr{NTuple{3,Int8}}) == 3
    @test Base.elsize(Ptr{Cvoid}) == 0
    @test Base.elsize(Ptr{Base.RefValue{Any}}) == sizeof(Int)
    @test Base.elsize(Ptr{Int}) == sizeof(Int)
```

vtjnash commented on test/intrinsics.jl:
```diff
@@ -107,16 +107,25 @@ end
+    @test Base.elsize(Ptr{Cvoid}) == 0
+    @test Base.elsize(Ptr{Base.RefValue{Any}}) == 8
+    @test Base.elsize(Ptr{Int}) == 8
+    @test_throws MethodError Base.elsize(Ptr)
```
```suggestion
    @test_throws MethodError Base.elsize(Ptr)
    @test_throws ErrorException Base.elsize(Ptr{Ref{Int}})
    @test_throws ErrorException Base.elsize(Ptr{Ref})
    @test_throws ErrorException Base.elsize(Ptr{Complex})
```

Tokazama committed:
Update test/intrinsics.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

Tokazama committed:
Update test/intrinsics.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

vtjnash commented on base/pointer.jl:
```diff
@@ -98,32 +98,147 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
+!!! compat "Julia 1.10"
+     This function requires at least Julia 1.10.
+
+See also: [`modifyproperty!`](@ref Base.modifyproperty!), [`atomic`](@ref)
```
```
┌ Error: no doc found for reference '[`Base.modifyproperty!`](@ref)' in src/base/c.md.
└ @ Documenter.CrossReferences /cache/build/default-amdci4-5/julialang/julia-master/doc/deps/packages/Documenter/yf96B/src/Utilities/Utilities.jl:32
┌ Error: no doc found for reference '[`Base.replaceproperty!`](@ref)' in src/base/c.md.
└ @ Documenter.CrossReferences /cache/build/default-amdci4-5/julialang/julia-master/doc/deps/packages/Documenter/yf96B/src/Utilities/Utilities.jl:32
┌ Error: no doc found for reference '[`Base.swapproperty!`](@ref)' in src/base/c.md.
└ @ Documenter.CrossReferences /cache/build/default-amdci4-5/julialang/julia-master/doc/deps/packages/Documenter/yf96B/src/Utilities/Utilities.jl:32
```

vtjnash replied:
(these need to be added to base/base.md alongside the existing getproperty one)

Tokazama committed:
Add atomic property and field methods to docs

Tokazama committed:
Fix mistake in resolving conflicts

akdienes commented on base/pointer.jl:
```diff
@@ -98,32 +98,147 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
 
 Load a value of type `T` from the address of the `i`th element (1-indexed) starting at `p`.
-This is equivalent to the C expression `p[i-1]`.
+This is equivalent to the C expression `p[i-1]`. Optionally, an atmoic memory ordering can
```
atmoic -> atomic

Tokazama replied:
Thanks!

Tokazama committed:
Fix spelling

Tokazama commented:
Not sure why tests are failing at this point. Docs should be good and only change to code was the implementation to elsize for pointers, which shouldn't be used anywhere else yet since it's new
